### ISSUE TITLE
Multicore training with Alexnet on fashion mnist (jupyter notebook tutorial) is very slow on google colab

### ISSUE THREAD

**Issue by KamranMostafavi** [pytorch/xla#3841]

**Comment by AlexWertheim** [comment]
Thanks for bringing this to our attention - I'll take a look. Could you please confirm that [this](https://github.com/pytorch/xla/blob/master/contrib/colab/multi-core-alexnet-fashion-mnist.ipynb) is the notebook to which you're referring?

**Comment by JackCaoG** [comment]
when you mentioned 30s, what's the GPU setup and how many epoch it runs?

**Comment by KamranMostafavi** [comment]
reply to AlexWertheim;
yes that seems to be it. Here is the link: (it is on pytorch/xla and I am assuming it is identical to yours)
https://colab.research.google.com/github/pytorch/xla/blob/master/contrib/colab/multi-core-alexnet-fashion-mnist.ipynb#scrollTo=SAzIwJg2vweI

**Comment by KamranMostafavi** [comment]
response to JackCaoG.
It is for TPU. I ran it on Colab with TPU as runtime type. It runs OK but per my bug report, it takes each TPU core more than 300s instead of expected 30s to complete its task.

**Comment by JackCaoG** [comment]
> response to JackCaoG. It is for TPU. I ran it on Colab with TPU as runtime type. It runs OK but per my bug report, it takes each TPU core more than 300s instead of expected 30s to complete its task.

right, I am wondering which 30s are you referring to. Did you try to run the AlexNext with 4v100 and saw it took30s to finish?

**Comment by KamranMostafavi** [comment]
Response to JackCaoG:
No I did not try to do this on GPU (4v100). 30 sec is what the text inside the Jupyter notebook says should take for each TPU to complete its task.  Here a cut/paste of it: 
"The network should take about 30 seconds to train and about 10 seconds to evaluate on each process. Using an entire Cloud TPU is, as expected, dramatically faster than training and evaluating on a single Cloud TPU core."

**Comment by KamranMostafavi** [comment]
Response to JackGaoG:
I ran a similar notebook on GPU and it trained each epoch in 48 seconds. Here is the link to notebook available on Kaggle
https://www.kaggle.com/code/tiiktak/fashion-mnist-with-alexnet-in-pytorch-92-accuracy/execution

**Comment by AlexWertheim** [comment]
Thanks very much for your data. I've reproduced your issue - right now, each core takes ~320 seconds on average to run. This is consistent across previous versions of PyTorch/XLA. The single core AlexNet example is also slower than expected, and takes ~540 seconds on average to run, which is again consistent across PyTorch/XLA versions. I'll try to understand the performance differences between GPU and TPU, and report back with what I find.

**Comment by AlexWertheim** [comment]
Thanks for your patience. At the moment, it appears that this is likely a colab issue, not a TPU or PyTorch/XLA issue. Running the same code natively on TPU VM, the speed estimates provided in the colab are pretty accurate; training takes ~40s per core, and training takes ~10s, as (more or less) advertised. 

All that is to say, the code seems to work as expected as long as you can run it on TPU VM directly. If I discover anything helpful on what's causing the performance differences with colab, I'll be sure to document them here.

**Comment by KamranMostafavi** [comment]
AlexWertheim,
Thanks for letting me know. I am wondering if you were able to reproduce the problem in Colab at your side? Or it is something related to my Colab account?  If it is not unique to my environment,  I also recommend to perhaps change the text in the notebook to indicate that it should be executed in TPU VM  as running it in Colab may not show the performance improvements over GPU.

**Comment by AlexWertheim** [comment]
Indeed, I was able to reproduce the issue on my side - see the comment preceding my most recent comment - so this isn't anything related to your Colab account. Thanks for the suggestion! Once we know a little more, we may change the Colab accordingly if need be.

**Comment by KamranMostafavi** [comment]
Excellent!  It seems then that this issue is pytorch/colab  dependent.  Do
you know if the same is true if the code was based on Keras/Tensor flow ?
that is TPU would not perform as expected? If you do not know, then does
XLA support Keras/Tensorflow or it is just pytorch?

**Comment by KamranMostafavi** [comment]
Just for comparison, I did try an implementation with Keras and it trained 1 epoch of MNIST for approx 23 sec with AlexNet as model on Google Colab with TPU.
